pkg/engine/plugins.go

```diff
@@ -183,3 +191,95 @@ func installPlugin(client deploy.BackendClient, plugin workspace.PluginInfo) err
 	logging.V(7).Infof("installPlugin(%s, %s): successfully installed", plugin.Name, plugin.Version)
 	return nil
 }
+
+// computeDefaultProviderPlugins computes, for every resource plugin, a mapping from packages to semver versions
+// reflecting the version of a provider that should be used as the "default" resource when registering resources. This
+// function takes two sets of plugins: a set of plugins given to us from the language host and the full set of plugins.
+// If the language host has sent us a non-empty set of plugins, we will use those exclusively to service default
+// provider requests. Otherwise, we will use the full set of plugins, which is the existing behavior today.
+//
+// The justification for favoring language plugins over all else is that, ultimately, it is the language plugin that
+// produces resource registrations and therefore it is the language plugin that should dictate exactly what plugins to
+// use to satisfy a resource registration. Since we do not today request a particular version of a plugin via
+// RegisterResource (pulumi/pulumi#2389), this is the best we can do to infer the version that the language plugin
+// actually wants.
+//
+// Whenever a resource arrives via RegisterResource and does not explicitly specify which provider to use, the engine
+// injects a "default" provider resource that will serve as that resource's provider. This function computes the map
+// that the engine uses to determine which version of a particular provider to load.
+//
+// it is critical that this function be 100% deterministic.
```

Comment:
any concern about sort stability?  do we ever potentially have the same version twice, but pointing at different acutal data?  should we consider error'ing in that sort of case?

Reply:
Our data structure does not have duplicate versions for a single plugin; we don't need to worry about stability.